idaes/core/util/tests/test_model_diagnostics.py

```diff
@@ -2883,3 +3037,224 @@ def test_display(self):
         self.assertEqual(len(exp_list), len(got_list))
         for _exp, _got in zip(exp_list, got_list):
             self.assertEqual(_exp, _got)
+
+
+class TestCheckParallelJacobian:
+    @pytest.mark.unit
+    def test_invalid_direction(self):
+        m = ConcreteModel()
+
+        with pytest.raises(
+            ValueError,
+            match="Unrecognised value for direction \(foo\). "
+            "Must be 'row' or 'column'.",
+        ):
+            check_parallel_jacobian(m, direction="foo")
+
+    @pytest.fixture(scope="class")
+    def model(self):
+        m = ConcreteModel()
+
+        m.v1 = Var(initialize=1e-8)
+        m.v2 = Var()
+        m.v3 = Var()
+        m.v4 = Var()
+
+        m.c1 = Constraint(expr=m.v1 == m.v2 - 0.99999 * m.v4)
+        m.c2 = Constraint(expr=m.v1 + 1.00001 * m.v4 == 1e-8 * m.v3)
+        m.c3 = Constraint(expr=1e8 * (m.v1 + m.v4) + 1e10 * m.v2 == 1e-6 * m.v3)
+        m.c4 = Constraint(expr=-m.v1 == -0.99999 * (m.v2 - m.v4))
+
+        return m
+
+    @pytest.mark.unit
+    def test_rows(self, model):
+        assert check_parallel_jacobian(model, direction="row") == [(model.c1, model.c4)]
+        assert check_parallel_jacobian(model, direction="row", tolerance=0) == []
+
+    @pytest.mark.unit
+    def test_columns(self, model):
+        assert check_parallel_jacobian(model, direction="column") == [
+            (model.v1, model.v4)
+        ]
+
+
+class TestCheckIllConditioning:
+    @pytest.mark.unit
+    def test_invalid_direction(self):
+        m = ConcreteModel()
+
+        with pytest.raises(
+            ValueError,
+            match="Unrecognised value for direction \(foo\). "
+            "Must be 'row' or 'column'.",
+        ):
+            check_ill_conditioning(m, direction="foo")
+
+    @pytest.fixture(scope="class")
+    def model(self):
+        m = ConcreteModel()
+
+        m.v1 = Var(initialize=1e-8)
+        m.v2 = Var()
+        m.v3 = Var()
+        m.v4 = Var()
+
+        m.c1 = Constraint(expr=m.v1 == m.v2 - 0.99999 * m.v4)
+        m.c2 = Constraint(expr=m.v1 + 1.00001 * m.v4 == 1e-8 * m.v3)
+        m.c3 = Constraint(expr=1e8 * (m.v1 + m.v4) + 1e10 * m.v2 == 1e-6 * m.v3)
+        m.c4 = Constraint(expr=-m.v1 == -0.99999 * (m.v2 - m.v4))
```

Comment:
Columns of the Jacobian corresponding to each variable:
```
v1: (1, 1, 1e8, -1)
v2: (-1, 0, 1e10, 1-1e5)
v3: (0, -1e8, -1e6, 0)
v4: (1-1e5, 1+1e5, 1e8, -(1-1e5))
```
`v1` and `v4` are parallel in an absolute sense, but we may also want to consider `(v1, v2)` and `(v2, v4)` to be parallel pairs in a relative sense, since 1 is very small relative to the large coefficients of `1e8` and `1e10`.

See my PR into this branch, where I've implemented this relative tolerance check.